Added geometry validation for GEO types to exit early on invalid latitudes

craigtaverner opened the pull request:
It turns out in #128234 that Lucene pushdown on ST_DISTANCE with invalid points (latitude out of range) will cause all documents to be returned, since DISJOINT on an invalid circle is true for all documents. We could either add an extra check for spatial pushdown that the geometries are valid, or add validation at geometry creation. This second option is much easier to implement, and a more comprehensive approach, as it prevents invalid geometries in many more places, hopefully reducing the likelihood of subtle and obscure bugs like #128234 happening in future.

Fixes #128234

craigtaverner committed:
Added geometry validation for GEO types to exit early on invalid latitudes

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

craigtaverner committed:
Update docs/changelog/128259.yaml

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

craigtaverner committed:
Merge branch 'main' into fix_invalid_geo

iverase reviewed (approved):
This change makes sense to me. I don't think is a breaking change but fixing a bug.

craigtaverner committed:
Fix failing tests

- ignore z-value in unit tests
- and use capabilities to block new tests on older versions that do not have this fix

craigtaverner committed:
Merge branch 'main' into fix_invalid_geo

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.19  | Commit could not be cherrypicked due to conflicts |
| ✅ |  [9.0](https://github.com/elastic/elasticsearch/pull/128425)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/128425">](https://github.com/elastic/elasticsearch/pull/128425) |
| ❌ |  8.17  | Commit could not be cherrypicked due to conflicts |
| ❌ |  8.18  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 128259`

craigtaverner commented:
Created manual backport at https://github.com/elastic/elasticsearch/pull/128426 for the 8.x branches, so removed the backport-pending label from this PR.
